File: src/imagedocumentloadpolicy.cpp
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

#include "imagedocumentloadpolicy.h"

#include "imagedocumentsourceloadrequest.h"
#include "imagedocumentstate.h"
#include "imagespreadpresentationcontroller.h"
#include "kiriview/src/imagedocumentloadpolicy.cxx.h"

namespace {
KiriView::RustImageDocumentSourceLoadKind rustSourceLoadKind(
    KiriView::ImageDocumentSourceLoadKind loadKind)
{
    switch (loadKind) {
    case KiriView::ImageDocumentSourceLoadKind::CurrentSource:
        return KiriView::RustImageDocumentSourceLoadKind::CurrentSource;
    case KiriView::ImageDocumentSourceLoadKind::ReplacementSource:
        return KiriView::RustImageDocumentSourceLoadKind::ReplacementSource;
    }

    return KiriView::RustImageDocumentSourceLoadKind::CurrentSource;
}

KiriView::RustImageDocumentRightToLeftReadingReset rustRightToLeftReadingReset(
    KiriView::ImageDocumentRightToLeftReadingReset reset)
{
    switch (reset) {
    case KiriView::ImageDocumentRightToLeftReadingReset::Keep:
        return KiriView::RustImageDocumentRightToLeftReadingReset::Keep;
    case KiriView::ImageDocumentRightToLeftReadingReset::ResetInactive:
        return KiriView::RustImageDocumentRightToLeftReadingReset::ResetInactive;
    case KiriView::ImageDocumentRightToLeftReadingReset::ResetActive:
        return KiriView::RustImageDocumentRightToLeftReadingReset::ResetActive;
    }

    return KiriView::RustImageDocumentRightToLeftReadingReset::Keep;
}

KiriView::RustImageDocumentSourceLoadPolicyInput rustSourceLoadPolicyInput(
    const KiriView::ImageDocumentSourceLoadPolicyInput &input)
{
    KiriView::RustImageDocumentSourceLoadPolicyInput rustInput {};
    rustInput.load_kind = rustSourceLoadKind(input.loadKind);
    rustInput.preserve_two_page_spread_transition = input.preserveTwoPageSpreadTransition;
    rustInput.right_to_left_reading_reset
        = rustRightToLeftReadingReset(input.rightToLeftReadingReset);
    rustInput.has_requested_container_navigation_url = input.hasRequestedContainerNavigationUrl;
    return rustInput;
}

KiriView::ImageDocumentRightToLeftReadingTransition rightToLeftReadingTransition(
    KiriView::RustImageDocumentRightToLeftReadingTransition transition)
{
    switch (transition) {
    case KiriView::RustImageDocumentRightToLeftReadingTransition::Keep:
        return KiriView::ImageDocumentRightToLeftReadingTransition::Keep;
    case KiriView::RustImageDocumentRightToLeftReadingTransition::Reset:
        return KiriView::ImageDocumentRightToLeftReadingTransition::Reset;
    case KiriView::RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyBeforeSourceState:
        return KiriView::ImageDocumentRightToLeftReadingTransition::ResetAndNotifyBeforeSourceState;
    case KiriView::RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyAfterOpen:
        return KiriView::ImageDocumentRightToLeftReadingTransition::ResetAndNotifyAfterOpen;
    }

    return KiriView::ImageDocumentRightToLeftReadingTransition::Keep;
}

KiriView::ImageDocumentSourceLoadUrlTarget sourceLoadUrlTarget(
    KiriView::RustImageDocumentSourceLoadUrlTarget target)
{
    switch (target) {
    case KiriView::RustImageDocumentSourceLoadUrlTarget::Unchanged:
        return KiriView::ImageDocumentSourceLoadUrlTarget::Unchanged;
    case KiriView::RustImageDocumentSourceLoadUrlTarget::Empty:
        return KiriView::ImageDocumentSourceLoadUrlTarget::Empty;
    case KiriView::RustImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation:
        return KiriView::ImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation;
    case KiriView::RustImageDocumentSourceLoadUrlTarget::RequestedSource:
        return KiriView::ImageDocumentSourceLoadUrlTarget::RequestedSource;
    }

    return KiriView::ImageDocumentSourceLoadUrlTarget::Unchanged;
}

KiriView::ImageDocumentSourceLoadPlan imageDocumentSourceLoadPlan(
    const KiriView::RustImageDocumentSourceLoadPlan &rustPlan)
{
    return KiriView::ImageDocumentSourceLoadPlan {
        rustPlan.cancel_navigation_and_predecode,
        rustPlan.finish_spread_transition,
        rightToLeftReadingTransition(rustPlan.right_to_left_reading_transition),
        rustPlan.clear_secondary_page,
        sourceLoadUrlTarget(rustPlan.loading_container_navigation_url),
        sourceLoadUrlTarget(rustPlan.container_navigation_url),
        sourceLoadUrlTarget(rustPlan.source_url),
        rustPlan.begin_open,
    };
}

KiriView::ImageDocumentSourceLoadKind sourceLoadKind(const KiriView::ImageDocumentState &state,
    const KiriView::ImageDocumentSourceLoadRequest &request)
{
    if (state.sourceUrl() == request.sourceUrl) {
        return KiriView::ImageDocumentSourceLoadKind::CurrentSource;
    }

    return KiriView::ImageDocumentSourceLoadKind::ReplacementSource;
}

KiriView::ImageDocumentRightToLeftReadingReset rightToLeftReadingReset(
    const KiriView::ImageDocumentState &state,
    const KiriView::ImageSpreadPresentationController &spreadController,
    const KiriView::ImageDocumentSourceLoadRequest &request)
{
    if (!spreadController.shouldResetRightToLeftReadingForLoad(
            state.displayedArchiveDocument(), request.sourceUrl, request.containerNavigationUrl)) {
        return KiriView::ImageDocumentRightToLeftReadingReset::Keep;
    }

    return spreadController.rightToLeftReadingEnabled()
        ? KiriView::ImageDocumentRightToLeftReadingReset::ResetActive
        : KiriView::ImageDocumentRightToLeftReadingReset::ResetInactive;
}

KiriView::ImageDocumentSourceLoadPolicyInput sourceLoadPolicyInput(
    const KiriView::ImageDocumentState &state,
    const KiriView::ImageSpreadPresentationController &spreadController,
    const KiriView::ImageDocumentSourceLoadRequest &request)
{
    KiriView::ImageDocumentSourceLoadPolicyInput input;
    input.loadKind = sourceLoadKind(state, request);
    input.preserveTwoPageSpreadTransition = request.preserveTwoPageSpreadTransition;
    input.rightToLeftReadingReset = rightToLeftReadingReset(state, spreadController, request);
    input.hasRequestedContainerNavigationUrl = !request.containerNavigationUrl.isEmpty();
    return input;
}
}

namespace KiriView::ImageDocumentLoadPolicy {
ImageDocumentSourceLoadPlan sourceLoadPlan(const ImageDocumentSourceLoadPolicyInput &input)
{
    return imageDocumentSourceLoadPlan(
        rustImageDocumentSourceLoadPlan(rustSourceLoadPolicyInput(input)));
}

ImageDocumentSourceLoadPlan sourceLoadPlan(const ImageDocumentState &state,
    const ImageSpreadPresentationController &spreadController,
    const ImageDocumentSourceLoadRequest &request)
{
    return sourceLoadPlan(::sourceLoadPolicyInput(state, spreadController, request));
}
}

File: src/imagedocumentloadpolicy.rs
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

#[cxx::bridge(namespace = "KiriView")]
mod ffi {
    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustImageDocumentSourceLoadKind {
        CurrentSource = 0,
        ReplacementSource = 1,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustImageDocumentRightToLeftReadingReset {
        Keep = 0,
        ResetInactive = 1,
        ResetActive = 2,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustImageDocumentRightToLeftReadingTransition {
        Keep = 0,
        Reset = 1,
        ResetAndNotifyBeforeSourceState = 2,
        ResetAndNotifyAfterOpen = 3,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustImageDocumentSourceLoadUrlTarget {
        Unchanged = 0,
        Empty = 1,
        RequestedContainerNavigation = 2,
        RequestedSource = 3,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustImageDocumentSourceLoadPolicyInput {
        load_kind: RustImageDocumentSourceLoadKind,
        preserve_two_page_spread_transition: bool,
        right_to_left_reading_reset: RustImageDocumentRightToLeftReadingReset,
        has_requested_container_navigation_url: bool,
    }

    #[derive(Debug, PartialEq, Eq)]
    struct RustImageDocumentSourceLoadPlan {
        cancel_navigation_and_predecode: bool,
        finish_spread_transition: bool,
        right_to_left_reading_transition: RustImageDocumentRightToLeftReadingTransition,
        clear_secondary_page: bool,
        loading_container_navigation_url: RustImageDocumentSourceLoadUrlTarget,
        container_navigation_url: RustImageDocumentSourceLoadUrlTarget,
        source_url: RustImageDocumentSourceLoadUrlTarget,
        begin_open: bool,
    }

    extern "Rust" {
        #[cxx_name = "rustImageDocumentSourceLoadPlan"]
        fn rust_image_document_source_load_plan(
            input: RustImageDocumentSourceLoadPolicyInput,
        ) -> RustImageDocumentSourceLoadPlan;
    }
}

use ffi::{
    RustImageDocumentRightToLeftReadingReset, RustImageDocumentRightToLeftReadingTransition,
    RustImageDocumentSourceLoadKind, RustImageDocumentSourceLoadPlan,
    RustImageDocumentSourceLoadPolicyInput, RustImageDocumentSourceLoadUrlTarget,
};

fn rust_image_document_source_load_plan(
    input: RustImageDocumentSourceLoadPolicyInput,
) -> RustImageDocumentSourceLoadPlan {
    if replaces_source(input) {
        return replacement_source_load_plan(input);
    }

    current_source_load_plan(input)
}

fn replaces_source(input: RustImageDocumentSourceLoadPolicyInput) -> bool {
    input.load_kind == RustImageDocumentSourceLoadKind::ReplacementSource
}

fn current_source_load_plan(
    input: RustImageDocumentSourceLoadPolicyInput,
) -> RustImageDocumentSourceLoadPlan {
    RustImageDocumentSourceLoadPlan {
        cancel_navigation_and_predecode: false,
        finish_spread_transition: !input.preserve_two_page_spread_transition,
        right_to_left_reading_transition: right_to_left_reading_transition(input, false),
        clear_secondary_page: false,
        loading_container_navigation_url: RustImageDocumentSourceLoadUrlTarget::Empty,
        container_navigation_url: if input.has_requested_container_navigation_url {
            RustImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation
        } else {
            RustImageDocumentSourceLoadUrlTarget::Unchanged
        },
        source_url: RustImageDocumentSourceLoadUrlTarget::Unchanged,
        begin_open: false,
    }
}

fn replacement_source_load_plan(
    input: RustImageDocumentSourceLoadPolicyInput,
) -> RustImageDocumentSourceLoadPlan {
    RustImageDocumentSourceLoadPlan {
        cancel_navigation_and_predecode: true,
        finish_spread_transition: !input.preserve_two_page_spread_transition,
        right_to_left_reading_transition: right_to_left_reading_transition(input, true),
        clear_secondary_page: true,
        loading_container_navigation_url:
            RustImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation,
        container_navigation_url: RustImageDocumentSourceLoadUrlTarget::Unchanged,
        source_url: RustImageDocumentSourceLoadUrlTarget::RequestedSource,
        begin_open: true,
    }
}

fn right_to_left_reading_transition(
    input: RustImageDocumentSourceLoadPolicyInput,
    replacement_source: bool,
) -> RustImageDocumentRightToLeftReadingTransition {
    match input.right_to_left_reading_reset {
        RustImageDocumentRightToLeftReadingReset::Keep => {
            RustImageDocumentRightToLeftReadingTransition::Keep
        }
        RustImageDocumentRightToLeftReadingReset::ResetInactive => {
            RustImageDocumentRightToLeftReadingTransition::Reset
        }
        RustImageDocumentRightToLeftReadingReset::ResetActive => {
            if replacement_source {
                RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyAfterOpen
            } else {
                RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyBeforeSourceState
            }
        }
        _ => RustImageDocumentRightToLeftReadingTransition::Keep,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn source_load_input(
        load_kind: RustImageDocumentSourceLoadKind,
        preserve_two_page_spread_transition: bool,
        right_to_left_reading_reset: RustImageDocumentRightToLeftReadingReset,
        has_requested_container_navigation_url: bool,
    ) -> RustImageDocumentSourceLoadPolicyInput {
        RustImageDocumentSourceLoadPolicyInput {
            load_kind,
            preserve_two_page_spread_transition,
            right_to_left_reading_reset,
            has_requested_container_navigation_url,
        }
    }

    fn current_source_plan(
        finish_spread_transition: bool,
        right_to_left_reading_transition: RustImageDocumentRightToLeftReadingTransition,
        container_navigation_url: RustImageDocumentSourceLoadUrlTarget,
    ) -> RustImageDocumentSourceLoadPlan {
        RustImageDocumentSourceLoadPlan {
            cancel_navigation_and_predecode: false,
            finish_spread_transition,
            right_to_left_reading_transition,
            clear_secondary_page: false,
            loading_container_navigation_url: RustImageDocumentSourceLoadUrlTarget::Empty,
            container_navigation_url,
            source_url: RustImageDocumentSourceLoadUrlTarget::Unchanged,
            begin_open: false,
        }
    }

    fn replacement_source_plan(
        finish_spread_transition: bool,
        right_to_left_reading_transition: RustImageDocumentRightToLeftReadingTransition,
    ) -> RustImageDocumentSourceLoadPlan {
        RustImageDocumentSourceLoadPlan {
            cancel_navigation_and_predecode: true,
            finish_spread_transition,
            right_to_left_reading_transition,
            clear_secondary_page: true,
            loading_container_navigation_url:
                RustImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation,
            container_navigation_url: RustImageDocumentSourceLoadUrlTarget::Unchanged,
            source_url: RustImageDocumentSourceLoadUrlTarget::RequestedSource,
            begin_open: true,
        }
    }

    #[test]
    fn source_load_plan_routes_unchanged_and_replacement_loads() {
        let unchanged = rust_image_document_source_load_plan(source_load_input(
            RustImageDocumentSourceLoadKind::CurrentSource,
            false,
            RustImageDocumentRightToLeftReadingReset::ResetActive,
            true,
        ));
        assert_eq!(
            unchanged,
            current_source_plan(
                true,
                RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyBeforeSourceState,
                RustImageDocumentSourceLoadUrlTarget::RequestedContainerNavigation,
            )
        );

        let replacement = rust_image_document_source_load_plan(source_load_input(
            RustImageDocumentSourceLoadKind::ReplacementSource,
            true,
            RustImageDocumentRightToLeftReadingReset::Keep,
            false,
        ));
        assert_eq!(
            replacement,
            replacement_source_plan(false, RustImageDocumentRightToLeftReadingTransition::Keep)
        );

        let inactive_reset_replacement = rust_image_document_source_load_plan(source_load_input(
            RustImageDocumentSourceLoadKind::ReplacementSource,
            true,
            RustImageDocumentRightToLeftReadingReset::ResetInactive,
            false,
        ));
        assert_eq!(
            inactive_reset_replacement,
            replacement_source_plan(false, RustImageDocumentRightToLeftReadingTransition::Reset)
        );

        let resetting_replacement = rust_image_document_source_load_plan(source_load_input(
            RustImageDocumentSourceLoadKind::ReplacementSource,
            true,
            RustImageDocumentRightToLeftReadingReset::ResetActive,
            false,
        ));
        assert_eq!(
            resetting_replacement,
            replacement_source_plan(
                false,
                RustImageDocumentRightToLeftReadingTransition::ResetAndNotifyAfterOpen
            )
        );
    }
}
